### PR TITLE
bpo-39037: Look up __enter__ before __exit__ in the with statement documentation

### PR DESCRIPTION
**Pull request by geryogam**
This PR will make these changes to the [Compound statements](https://docs.python.org/3/reference/compound_stmts.html) chapter of the language documentation:

- fix the trial order of the `__exit__` and `__enter__` methods in the `with` statement;
- add the equivalent `try` statement of the `with` statement (since it is already given for the `async with` statement);
- replace the `else` clause with a `finally` clause in the equivalent `try` statement of the `async with` statement, in order to also handle the case when there is a non-local goto statement (`break`, `continue`, `return`) in the suite (cf. https://stackoverflow.com/questions/59322585/what-is-the-exact-try-statement-equivalent-of-the-with-statement-in-python).

https://bugs.python.org/issue39037

### CODE REVIEW

File: Doc/reference/compound_stmts.rst [hunk 1]
```diff
@@ -399,10 +399,10 @@ The execution of the :keyword:`with` statement with one "item" proceeds as follo
 #. The context expression (the expression given in the :token:`with_item`) is
    evaluated to obtain a context manager.
 
-#. The context manager's :meth:`__exit__` is loaded for later use.
-
 #. The context manager's :meth:`__enter__` method is invoked.
```
Comment (serhiy-storchaka) [inline Doc/reference/compound_stmts.rst]: The right sequence is:
* The `__enter__` method is loaded.
* The `__exit__` method is loaded.
* The `__enter__` method is invoked.

File: Doc/reference/compound_stmts.rst [hunk 2]
```diff
@@ -430,13 +430,36 @@ The execution of the :keyword:`with` statement with one "item" proceeds as follo
...
 
+The following code::
+
+    with expression as target:
+        suite
+
+is semantically equivalent to::
+
+    manager = (expression)
+    value = type(manager).__enter__(manager)
```
Comment (serhiy-storchaka) [inline Doc/reference/compound_stmts.rst]: It is called after getting `__exit__`.

File: Doc/reference/compound_stmts.rst [hunk 3]
```diff
@@ -811,23 +834,26 @@ able to suspend execution in its *enter* and *exit* methods.
 
 The following code::
 
-    async with EXPR as VAR:
-        BLOCK
+    async with expression as target:
```
Comment (serhiy-storchaka) [inline Doc/reference/compound_stmts.rst]: What is the purpose of this change?

File: Doc/reference/compound_stmts.rst [hunk 4]
```diff
@@ -811,23 +834,26 @@ able to suspend execution in its *enter* and *exit* methods.
...
-        BLOCK
+        suite
     except:
-        if not await aexit(mgr, *sys.exc_info()):
+        exception = True
+        if not await aexit(manager, *sys.exc_info()):
             raise
-    else:
-        await aexit(mgr, None, None, None)
+    finally:
```
Comment (serhiy-storchaka) [inline Doc/reference/compound_stmts.rst:859]: Why replace `else` with `finally`?

Comment (maggyero) [reply Doc/reference/compound_stmts.rst]: In order to also handle the case when there is a non-local goto statement (`break`, `continue`, `return`) in the suite, cf. https://www.python.org/dev/peps/pep-0343/#specification-the-with-statement.

Comment (geryogam) [reply Doc/reference/compound_stmts.rst:859]: @serhiy-storchaka  To handle the case when there is a non-local goto statement (`break`, `continue`, `return`) in the suite (cf. https://www.python.org/dev/peps/pep-0343/#specification-the-with-statement). Before knowing about PEP 343 I did the same mistake (cf. https://stackoverflow.com/questions/59322585/what-is-the-exact-try-statement-equivalent-of-the-with-statement-in-python).

File: Doc/reference/compound_stmts.rst (hunk 3 above)
Comment (geryogam) [reply Doc/reference/compound_stmts.rst]: @serhiy-storchaka To use the same terms as in the grammar.

File: Doc/reference/compound_stmts.rst (hunk 1 above)
Comment (geryogam) [reply Doc/reference/compound_stmts.rst]: @serhiy-storchaka You are right, I am going to update this, thanks.

File: Doc/reference/compound_stmts.rst (hunk 2 above)
Comment (geryogam) [reply Doc/reference/compound_stmts.rst]: @serhiy-storchaka True, thanks.

File: Doc/reference/compound_stmts.rst (hunk 4 above)
Comment (maggyero) [reply Doc/reference/compound_stmts.rst]: But before knowing about PEP 343 I did the same mistake: https://stackoverflow.com/questions/59322585/what-is-the-exact-try-statement-equivalent-of-the-with-statement-in-python

Comment (maggyero) [reply Doc/reference/compound_stmts.rst]: @serhiy-storchaka  But before knowing about PEP 343 I did the same mistake: https://stackoverflow.com/questions/59322585/what-is-the-exact-try-statement-equivalent-of-the-with-statement-in-python

File: Doc/reference/compound_stmts.rst (hunk 3 above)
Comment (brettcannon) [reply Doc/reference/compound_stmts.rst]: It's not important to match the grammar in this case as the docs are not kept in sync with the grammar rule names.

**Comment by brettcannon** [comment]
@maggyero your PR is failing CI.

Adding @ncoghlan as a reviewer since he probably is a good judge about the accuracy of the proposed clarification.

**Comment by maggyero** [comment]
@brettcannon Fixed, CI passed, thanks.

File: Doc/reference/compound_stmts.rst (hunk 3 above)
Comment (ncoghlan) [reply Doc/reference/compound_stmts.rst]: Right, we deliberately use ALL_CAPS in these semantic expansions to distinguish the placeholders from keywords and hidden variables.

Comment (geryogam) [reply Doc/reference/compound_stmts.rst]: @ncoghlan All right, updated.

**Review by ncoghlan — APPROVED** [review]

**Comment by maggyero** [comment]
Thanks @serhiy-storchaka, @brettcannon and @ncoghlan for reviewing this PR!
